ext/TensorOperationscuTENSORExt.jl

```diff
@@ -1,350 +1,306 @@
     end
-
-    return C
+    @eval function TO.tensorcontract!(C::$ArrayType,
+                                      A::$ArrayType, pA::Index2Tuple, conjA::Symbol,
+                                      B::$ArrayType, pB::Index2Tuple, conjB::Symbol,
+                                      pAB::Index2Tuple, α::Number, β::Number)
+        return tensorcontract!(C, A, pA, conjA, B, pB, conjB, pAB, α, β, cuTENSORBackend())
+    end
+    @eval function TO.tensortrace!(C::$ArrayType,
+                                   A::$ArrayType, p::Index2Tuple, q::Index2Tuple,
+                                   conjA::Symbol,
+                                   α::Number, β::Number)
+        return tensortrace!(C, A, p, q, conjA, α, β, cuTENSORBackend())
+    end
+    @eval function TO.tensoradd_type(TC, ::$ArrayType, pA::Index2Tuple, conjA::Symbol)
+        return CUDA.CuArray{TC,TO.numind(pA)}
+    end
+    @eval function TO.tensorcontract_type(TC,
+                                          ::$ArrayType, pA::Index2Tuple, conjA::Symbol,
+                                          ::$ArrayType, pB::Index2Tuple, conjB::Symbol,
+                                          pAB::Index2Tuple)
+        return CUDA.CuArray{TC,TO.numind(pAB)}
+    end
+    @eval TO.tensorfree!(C::$ArrayType) = TO.tensorfree!(C::$ArrayType, cuTENSORBackend())
 end
 
-function TO.tensorcontract!(C::AnyCuArray, pC::Index2Tuple,
-                            A::AnyCuArray, pA::Index2Tuple, conjA::Symbol,
-                            B::AnyCuArray, pB::Index2Tuple, conjB::Symbol,
-                            α::Number, β::Number)
-    return tensorcontract!(C, pC, A, pA, conjA, B, pB, conjB, α, β, StridedCUDA())
+# making sure that if the backend is specified, arrays are converted to CuArrays
+
+function TO.tensoradd!(C::AbstractArray,
+                       A::AbstractArray, pA::Index2Tuple, conjA::Symbol,
+                       α::Number, β::Number,
+                       backend::cuTENSORBackend)
+    C_cuda = adapt(CuArray, C)
+    A_cuda = adapt(CuArray, A)
+    tensoradd!(C_cuda, A_cuda, pA, conjA, α, β, backend)
+    C === C_cuda || copyto!(C, C_cuda)
+    return C
 end
-function TO.tensorcontract!(C::AbstractArray, pC::Index2Tuple,
+function TO.tensorcontract!(C::AbstractArray,
                             A::AbstractArray, pA::Index2Tuple, conjA::Symbol,
                             B::AbstractArray, pB::Index2Tuple, conjB::Symbol,
-                            α::Number, β::Number, backend::StridedCUDA)
-    tensorcontract!(StridedView(C), pC, StridedView(A), pA, conjA,
-                    StridedView(B), pB, conjB, α, β, backend)
+                            pAB::Index2Tuple,
+                            α::Number, β::Number, backend::cuTENSORBackend)
+    C_cuda = adapt(CuArray, C)
+    A_cuda = adapt(CuArray, A)
+    B_cuda = adapt(CuArray, B)
+    tensorcontract!(C_cuda, A_cuda, pA, conjA, B_cuda, pB, conjB, pAB, α, β, backend)
+    C === C_cuda || copyto!(C, C_cuda)
```

Comment:
I would prefer `copy!` over `copyto!`. The former is more strict in requiring equal size (although they probably amount to the same and have the same efficiency).

Reply:
Ah, I forgot that method exists. There are too many variants of `copy`...